### PR TITLE
Expose 'buttonClass' input to set custom classes to all buttons

### PR DESCRIPTION
**Pull request by MurhafSousli**
- Add `buttonClass` input to `<share-buttons>` component: sets custom classes to all buttons

   _before it was only possible to set custom classes in a single button `<share-button>`
   now it's possible in  `<share-buttons>`_ 

 - Improve folder structure
 - Move static functions to Helper `share.helper.ts`
 - Remove duplicate code:
   - use nFormatter in the pipe from Helper
   - use error handling function from Helper

Possible TODO: convert `ShareArgs` and `ShareButton` class to interfaces (low priority), @tinesoft  what do you think?

I was also trying to make integration possible with external button components like material, I thought about adding a new input to specify the name of the directive that should be added to all buttons, e.g. `md-button`, but It seems NOT possible to do that dynamically! 
if it worth it, we should find a way without using the `*ngIf`

Cheers!

### CODE REVIEW

**Comment by tinesoft** [comment]
Hi Murhaf,

Thanks for considering my humble opinions for future evolution of the library. I appreciate 👍 

A couple of suggestions before i continue, to improve future collaboration and review:
- **commits**: I recommend following  [Angular Git Commit Guidelines](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit) as i do. 
They really improve readability of commits messages, and somehow enforce unit, single-focused commits 
- **commits**:  Keep them focused and related (avoid doing anything out of the mentioned scope). Several separate  commits are better/easier to review than one big all-in commit.

Now regarding your questions, 

- **converting ShareArgs and ShareButton to interfaces**: given that those are only used internally, it make no much sense to transform them, as final users will not need to extend them.
- **integration  with external button components**:  that could be great.  I suggest delaying the discussion about it until i add tests for current components, as this will certainly trigger huge API refactoring.

Regarding this commit, i will comment on the PR directly (next to related source code changes).

I will post back here, once i'm done. So keep this commit as is for the moment

File: .gitignore [hunk 1]
```diff
@@ -4,3 +4,4 @@ dist
 coverage
 npm-debug.log
 
+/demo/src/app/share/
```
Comment (tinesoft) [inline .gitignore]: With my PR #32, i think all changes made to demo project  in this commit, can be reverted.

File: package.json [hunk 2]
```diff
@@ -49,7 +49,7 @@
     "angular2-template-loader": "^0.6.0",
     "autoprefixer": "^6.6.1",
     "awesome-typescript-loader": "^2.2.3",
-    "codelyzer": "^1.0.0-beta.0",
+    "codelyzer": "^1.0.0-beta.4",
```
Comment (tinesoft) [inline package.json]: Unrelated to this commit! should be done in separate commit

File: src/components/share-button/share-button.component.ts [hunk 3]
```diff
@@ -5,21 +5,19 @@ import {
...
+import {ShareButton, ShareArgs} from '../../classes/share-buttons.class';
+import {ShareButtonsService} from '../../services/share-buttons.service';
+import {WindowService} from '../../services/window.service';
+import {ShareProvider} from '../../helpers/share-provider.enum';
+import {Helper} from '../../helpers/share.helper';
 
 @Component({
     selector: 'share-button',
-    template: `<button  #btn (click)='share()'></button>`,
+    template: ``,
```
Comment (tinesoft) [inline src/components/share-button/share-button.component.ts]: Wrong move IMHO! 👎  
I think we should avoid DOM manipulation as much as possible and instead privilege the  Angular Way (using Templates, @ViewChild, ..) which is recommended and offer the advantage of being statically analyzable (AoT) 
for earlier detection of errors.

File: src/components/share-button/share-button.component.ts [hunk 4]
```diff
@@ -37,64 +35,34 @@ export class ShareButtonComponent implements AfterViewInit {
...
-
-        /** Set button template */
-        this.renderer.setElementProperty(this.btn.nativeElement, 'innerHTML', this.button.template);
-
-        /** Set buttons classes */
-        let classes = this.button.classes.match(/\S+/g) || [];
-        classes.map((btnClass) => this.renderer.setElementClass(this.btn.nativeElement, btnClass, true));
-
+        /** Validate URL */
+        this.url = Helper.validateUrl(this.url, this.window);
```
Comment (tinesoft) [inline src/components/share-button/share-button.component.ts]: Moving validation logic to `Helper.validateUrl() ` is the right think to do 👍 
But instead of passing  the entire window object to `validateUrl()`, which would make the method harder to test 
I suggest keep the default location retrieval in the component and update the method as such:

`validateUrl(url: string, defaultUrl?:string)`

UI related code should be keept in components and not be move to services

File: src/helpers/n-formatter.pipe.ts [hunk 5]
```diff
@@ -1,32 +0,0 @@
-import { Pipe, PipeTransform } from '@angular/core';
```
Comment (tinesoft) [inline src/helpers/n-formatter.pipe.ts]: Following my comment about **avoiding  DOM manipulation** (see above), removing the pipe is not good, as ideally it could be used in the template for share-button

File: demo/src/app/share/helpers/share.helper.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+                userId: '@viewer',
+                groupId: '@self'
+            },
+            jsonrpc: '2.0',
+            key: 'p',
+            apiVersion: 'v1'
+        }];
+    };
+
+    export const handleError = (err) => {
```
Comment (tinesoft) [inline demo/src/app/share/helpers/share.helper.ts]: Does this really make sense outside the `share-buttons.service.ts` ?
You are returning an empty observable, which is strongly linked to implementation of ..count methods...

I think this should be kept in the service.

File: src/services/share-buttons.service.ts [hunk 7]
```diff
@@ -1,39 +1,18 @@
...
+import { Observable } from 'rxjs/Observable';
 
+import { ShareArgs } from '../classes/share-buttons.class';
+import { ShareProvider } from '../helpers/share-provider.enum';
+import { ShareLinks } from '../helpers/share-links.functions';
+import { Helper } from '../helpers/share.helper';
 
 @Injectable()
-export class ShareButtonsService implements ShareButtonsInterface {
+export class ShareButtonsService {
```
Comment (tinesoft) [inline src/services/share-buttons.service.ts]: Unrelated to this commit! should be done in separate commit (eg: 'refactoring of ShareButtonsService')

File: src/components/share-button/share-button.component.ts [hunk 8]
```diff
@@ -105,24 +73,25 @@ export class ShareButtonComponent implements AfterViewInit {
...
-            if (num >= si[i].value) {
-                return (num / si[i].value).toFixed(digits).replace(rx, '$1') + si[i].symbol;
-            }
-        }
-        return num.toFixed(digits).replace(rx, '$1');
+    shareCount() {
+        this.sbService.count(this.button.provider, this.url)
+            .subscribe(shareCount => {
+                if (shareCount) {
+                    let counter = this.renderer.createElement(this.self, 'span');
```
Comment (tinesoft) [inline src/components/share-button/share-button.component.ts]: Following the **"avoid DOM manipulation"** comment, this should be in the template of `share-button`,  and controlled by a `ngIf` on public field `shareCount`

File: src/components/share-button/share-button.component.ts [hunk 9]
```diff
@@ -105,24 +73,25 @@ export class ShareButtonComponent implements AfterViewInit {
                 this.popUpClosed.emit(this.button.provider);
             }
         }, 200);
-    }
+    };
 
+    shareButton() {
```
Comment (tinesoft) [inline src/components/share-button/share-button.component.ts]: Following the **"avoid DOM manipulation"** comment, this should be in the template of `share-button`

File: src/components/share-buttons/share-buttons.component.ts [hunk 10]
```diff
@@ -38,6 +38,9 @@ export class ShareButtonsComponent implements OnInit {
     /** Indicates weather default style is applied to the buttons */
     @Input() defaultStyle: boolean = true;
 
+    /** Add default class to all buttons */
+    @Input() buttonClass: string = '';
+
```
Comment (tinesoft) [inline src/components/share-buttons/share-buttons.component.ts]: Changes in this file are truly the only ones that  should be kept in this commit, which is about "**Exposing  'buttonClass' input to set custom classes to all buttons**"

Other changes:
- improving folder structure by moving files (pipes, classes, styles)
- refactoring of ShareButtonsService
-...

should be in separate consecutive commits (ideally on **different  git branches**, to ease merging or cherry-picking of commits)

File: src/helpers/share.helper.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+            } else {
+                console.warn('ShareButtons: Invalid URL, switching to window.location.href');
+            }
+        }
+        /** fix window on Universal */
+        return (window) ? encodeURIComponent(window.location.href) :
+            (typeof global !== 'undefined') ? (<any>global).url : '';
+    };
+
+    export const nFormatter = (num, digits) => {
```
Comment (tinesoft) [inline src/helpers/share.helper.ts]: Does this really make sense outside the `n-formatter.pipe.ts` ?

File: src/helpers/share.helper.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        for (i = 0; i < si.length; i++) {
+            if (num >= si[i].value) {
+                return (num / si[i].value).toFixed(digits).replace(rx, '$1') + si[i].symbol;
+            }
+        }
+        return num.toFixed(digits).replace(rx, '$1');
+    };
+
+    /** Prepare gPlus count request body   */
+    export const gplusCountBody = (url) => {
```
Comment (tinesoft) [inline src/helpers/share.helper.ts]: Does this really make sense outside the `share-buttons.service.ts` ?
This is strongly linked to implementation of `gPlusCount()` method...

I think this should be kept in the service.

File: src/helpers/share.helper.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+                userId: '@viewer',
+                groupId: '@self'
+            },
+            jsonrpc: '2.0',
+            key: 'p',
+            apiVersion: 'v1'
+        }];
+    };
+
+    export const handleError = (err) => {
```
Comment (tinesoft) [inline src/helpers/share.helper.ts]: Does this really make sense outside the `share-buttons.service.ts` ?
You are returning an empty observable, which is strongly linked to implementation of ..count methods...

I think this should be kept in the service.

File: src/components/share-button/share-button.component.ts (hunk 8 above)
Comment (MurhafSousli) [inline src/components/share-button/share-button.component.ts]: Well, The plan was like this from the beginning, I tried 3 different ways before using `renderer`

 - if you use `async` pipe in the view it will send +1000 requests and will get the blocked.

 - If you make a child component e.g. `<share-count>` and use `[count]=sbService.count(button.provider, url) | async` you will also get the same result

 - if you assign the count result value to a variable and bind it to the view, then you have to remove the `ChangeDetectionStrategy.OnPush` because it won't update

By using the `renderer` I was able to add the count element only when the count is enabled, which gave the expected behavior, this was the only solution I found, other solution could be using a single state for the whole package and this requires some big changes.

So since I used the `renderer` once I made the changes of the `"avoid DOM manipulation" comment` to make it one style you know...

On the other hand I do not have experience in writing tests yet, so I'm not aware of the difficulties on that area

Have you been able to get it work using the template?

File: src/helpers/share.helper.ts (hunk 11 above)
Comment (MurhafSousli) [inline src/helpers/share.helper.ts]: Yupp, this function was written twice if you have noticed, one in the pipe and one in share-button component, if we are able to get the count in the sharebutton template (without the renderer) then we can move it back to the pipe

File: src/components/share-button/share-button.component.ts (hunk 8 above)
Comment (tinesoft) [inline src/components/share-button/share-button.component.ts]: Hi, 

I wasn't aware of theses issues, in particular with the 3rd option.
If that was the only reason for manipulating the DOM,  i think we can go full template and still benefit from change detection, by using [ChangeDetectorRef.markForCheck()](https://angular.io/docs/ts/latest/api/core/index/ChangeDetectorRef-class.html).  :)

See [this interesting article](https://blog.thoughtram.io/angular/2016/02/22/angular-2-change-detection-explained.html) by **Pascal Precht**.

Regarding tests, the problem with `ElementRef`, is that it cannot be directly injected in test environment, i have to do ugly things like creating a Mock class that extends ElementRef, and used it like this:

```
class MockElementRef extends ElementRef{}

  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [HttpModule, JsonpModule],
      declarations: [ShareButtonComponent],
      providers: [WindowService, ShareButtonsService, Renderer, { provide: ElementRef, useClass: MockElementRef }]
    })
      .compileComponents();
  }));
```

File: src/helpers/share.helper.ts (hunk 11 above)
Comment (tinesoft) [inline src/helpers/share.helper.ts]: Yes, that what i meant! ( I was already considering the 'template' solution :))

File: src/helpers/share.helper.ts (hunk 12 above)
Comment (MurhafSousli) [inline src/helpers/share.helper.ts]: @tinesoft I usually take out all static function to the helper module, and why would you keep it in the service if it's static? 
`share.helper.ts` is meant to include all static helper functions
Specially this one is not in harmony with the others, it's odd and just for gPlus post request body but we still have `gPlusCount()` in the service
Even Webstorm hates it, I think this makes the file shorter, cleaner easier to read, and we shouldn't have more than _200LOC_ per file

Comment (tinesoft) [inline src/helpers/share.helper.ts]: That a valid point 👍

File: src/helpers/share.helper.ts (hunk 13 above)
Comment (MurhafSousli) [inline src/helpers/share.helper.ts]: It doesn't matter for me really, but in order be DRY we should have taken the 2 lines from each catch and put them into one function

File: src/components/share-button/share-button.component.ts (hunk 8 above)
Comment (MurhafSousli) [inline src/components/share-button/share-button.component.ts]: Cool, I didn't know about `ChangeDetectorRef.markForCheck` before, this is exactly what we need
